q2_demux/_demux.py

```diff
@@ -1,54 +1,123 @@
+import os.path
 import gzip
 import yaml
 import itertools
 import collections
 
 import skbio
+import pandas as pd
+import seaborn as sns
 
 import qiime
 from q2_types.per_sample_sequences import (
-    SingleLanePerSampleSingleEndFastqDirFmt, FastqManifestFormat, YamlFormat)
+    SingleLanePerSampleSingleEndFastqDirFmt, FastqManifestFormat, YamlFormat,
+    FastqGzFormat)
+
+
+def _read_fastq_seqs(filepath):
+    # This function is adapted from @jairideout's SO post:
+    # http://stackoverflow.com/a/39302117/3424666
+    fh = gzip.open(filepath, 'rt')
+    for seq_header, seq, qual_header, qual in itertools.zip_longest(*[fh] * 4):
+        yield (seq_header.strip(), seq.strip(), qual_header.strip(),
+               qual.strip())
 
 
 def _extract_common_header(header):
     return header.rsplit('/', 1)[0]
 
 
-class BarcodeSequenceIterator(collections.Iterator):
+class BarcodeSequenceFastqIterator(collections.Iterator):
     def __init__(self, barcode_generator, sequence_generator):
         self.barcode_generator = barcode_generator
         self.sequence_generator = sequence_generator
 
     def __iter__(self):
         # Adapted from q2-types
-        for barcode, sequence in itertools.zip_longest(
+        for barcode_fields, sequence_fields in itertools.zip_longest(
                 self.barcode_generator, self.sequence_generator):
-            if barcode is None:
+            if barcode_fields is None:
                 raise ValueError('More sequences were provided than barcodes.')
-            if sequence is None:
+            if sequence_fields is None:
                 raise ValueError('More barcodes were provided than sequences.')
             # The id or description fields may end with "/read-number", which
-            # will differ between the sequence and barcode reads. Confirm that
-            # they are identical up until the last /
-            if _extract_common_header(barcode.metadata['id']) != \
-               _extract_common_header(sequence.metadata['id']):
+            # and read number will differ between the sequence and barcode
+            # reads. Confirm that they are identical up until the last /
+            barcode_header_fields = \
```

Comment:
define the namedtuple class in the module instead of within the function. i'd put it towards the top of the module, below the imports